docs: [IAC-2021]: Document Terraform backend initialization (#6899)
harness/developer-hub: docs/infra-as-code-management/terraform-backend/init-configuration.md

:::

## Apply environment variables
Using environment variables allows for a more dynamic and flexible backend configuration. By specifying an empty or partially configured backend block, we can populate the necessary values through environment variables.

### Partial configuration example
```hcl
terraform {
  backend "s3" {}
}
```

In this example, the backend configuration is left empty, allowing us to supply the values using environment variables during the `terraform init` command.

## Set environment variables
You can set the environment variables used to configure the backend in various ways, such as directly in the terminal or through Infrastructure pipelines. Here are the variables corresponding to the partial configuration:

:::important
To set up an environment variable that populates the terraform backend configuration as part of the `terraform init -backend-config` command (for example: `terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`), you must name your variables in the following format:

> 
> 

Reviewer: ```suggestion
```

      webkitallowfullscreen="webkitallowfullscreen" 
      mozallowfullscreen="mozallowfullscreen" 
      allowfullscreen="allowfullscreen"></iframe>
  </TabItem>
  <TabItem value="Step-by-step">
    1. Navigate to the **Infrastructure** module.
    2. Select **Workspaces**, then select the **Variables** tab.
    3. Select **New Environment Variable**.
    4. Add your variables:
        - **Key**: PLUGIN_INIT_BACKEND_CONFIG_*IDENTIFIER*
        - **Value**: Parameter `key=value` pair, e.g. `region=us-east-1`

  ![Terraform init environment variables](./static/tf-init-env-variables.png)
  </TabItem>
</Tabs>

## Review variable usage

Once your Terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.  

![Pipeline environment variable usage](./static/tf-initi-variable-log.png)

Notice the list of backend config values, shortly followed by the generated `terraform init` command including the `-backend-config` label with accompanying parameters, populated with your environment variable values.

> 

Reviewer: ```suggestion
```

---
title: Initialization
description: Initialize your infrastructure dynamically with environment variables
sidebar_position: 10
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Dynamic Backend Configuration

Reviewer: ```suggestion
# Dynamic backend configuration
```

At the first step of most infrastructure provisioning pipelines, you prepare your working directory with the `terraform init` command. Part of this step includes setting up your backend environment for Terraform state management.

Harness enables the use of environment variables to dynamically update your backend configuration in Terraform. This approach offers several benefits, including increased flexibility, easier configuration management, and improved security by avoiding hardcoded values.

## Hardcoded backend configuration
Hardcoding backend configuration in Terraform is straightforward but static. Here is an example of a hardcoded backend configuration:

```hcl
terraform {
  backend "s3" {
    bucket = "mybucket"
    key    = "path/to/my/key"
    region = "us-east-1"
  }
}
```

:::tip

> While this approach works, it lacks flexibility and can be challenging to update or manage in different environments (e.g., development, staging, production).

Reviewer: ```suggestion
While this approach works, it lacks flexibility and can be challenging to update or manage in different environments (for example, development, staging, production).
```

:::

## Apply environment variables
Using environment variables allows for a more dynamic and flexible backend configuration. By specifying an empty or partially configured backend block, we can populate the necessary values through environment variables.

> ### Partial Configuration Example

Reviewer: ```suggestion
### Partial configuration example
```

> ### Partial Configuration Example

Reviewer: Recommend using sentence case for all headings.

https://harness.atlassian.net/wiki/spaces/DOCS/pages/21544468526/Harness+documentation+style+guide#Headings%2C-titles

```hcl
terraform {
  backend "s3" {}
}
```

In this example, the backend configuration is left empty, allowing us to supply the values using environment variables during the `terraform init` command.

> ## Setting Environment Variables

Reviewer: ```suggestion
## Set environment variables
```

> ## Setting Environment Variables

Reviewer: Recommend avoiding gerunds for all headings.

https://harness.atlassian.net/wiki/spaces/DOCS/pages/21544468526/Harness+documentation+style+guide#Headings%2C-titles

> The environment variables used to configure the backend can be set in various ways, such as directly in the terminal or through CI/CD pipelines. Here are the variables corresponding to the partial configuration:

Reviewer: ```suggestion
You can set the environment variables used to configure the backend in various ways, such as directly in the terminal or through CI/CD pipelines. Here are the variables corresponding to the partial configuration:
```

:::important

> To set up an environment variable that population the terraform backend configuration as part of the `terraform init -backend-config` command, for example:

Reviewer: ```suggestion
To set up an environment variable that populates the terraform backend configuration as part of the `terraform init -backend-config` command (for example: `terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`), you must name your variables in the following format:
```

> `terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`

Reviewer: ```suggestion
```

> 
> You must name your variables in the following format:

Reviewer: ```suggestion
```

`PLUGIN_PLUGIN_INIT_BACKEND_CONFIG_` followed by a unique identifier to distinguish your variable.
:::

> When initializing an AWS S3 bucket for example, and passing the `bucket`, `key` and `region` parameters, Harness recommends naming your variables similar to the following:

Reviewer: ```suggestion
When initializing an AWS S3 bucket for example, and passing the `bucket`, `key`, and `region` parameters, Harness recommends naming your variables similar to the following:
```

- `PLUGIN_INIT_BACKEND_CONFIG_BUCKET`
- `PLUGIN_INIT_BACKEND_CONFIG_KEY`
- `PLUGIN_INIT_BACKEND_CONFIG_REGION`

> ## Adding New Environment Variables

Reviewer: ```suggestion
## Add new environment variables
```

To add new environment variables, follow these steps:

<Tabs>
<TabItem value="Interactive guide">
    <iframe 
      src="https://app.tango.us/app/embed/aa30519d-534a-4017-a0bb-8d0439f5258a" 
      title="Configure your Terraform backend config with Harness environment variables" 
      style={{minHeight:'640px'}}
      width="100%" 
      height="100%" 
      referrerpolicy="strict-origin-when-cross-origin" 
      frameborder="0" 
      webkitallowfullscreen="webkitallowfullscreen" 
      mozallowfullscreen="mozallowfullscreen" 
      allowfullscreen="allowfullscreen"></iframe>
  </TabItem>
  <TabItem value="Step-by-step">
    1. Navigate to the **Infrastructure** module.
    2. Select **Workspaces**, then select the **Variables** tab.
    3. Select **New Environment Variable**.
    4. Add your variables:
        - **Key**: PLUGIN_INIT_BACKEND_CONFIG_*IDENTIFIER*
        - **Value**: Parameter `key=value` pair, e.g. `region=us-east-1`

  ![Terraform init environment variables](./static/tf-init-env-variables.png)
  </TabItem>
</Tabs>

## Review variable usage

> Once your terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.

Reviewer: ```suggestion
Once your Terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.
```

![Pipeline environment variable usage](./static/tf-initi-variable-log.png)

Notice the list of backend config values, shortly followed by the generated `terraform init` command including the `-backend-config` label with accompanying parameters, populated with your environment variable values.

## Conclusion

> Using environment variables for backend configuration in Terraform provides a flexible and dynamic approach, making it easier to manage different environments and improve security. By following the steps outlined in this document, you can set up and use environment variables to configure your Terraform backend dynamically.

Reviewer: ```suggestion
Using environment variables for backend configuration in Terraform provides a flexible and dynamic approach, making it easier to manage different environments and improve security. By following the steps outlined in this topic, you can set up and use environment variables to configure your Terraform backend dynamically.
```
